### PR TITLE
[Merged by Bors] - feat: add `List.iterate_add`

### PR DESCRIPTION
**Pull request by eric-wieser**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Iterate.lean [hunk 1]
```diff
@@ -17,6 +17,14 @@ variable {α : Type*}
 
 namespace List
 
+theorem take_range (m n : ℕ) : take m (range n) = range (min m n) := by
+  apply List.ext_get
+  · simp
+  · simp only [length_take, length_range, lt_min_iff, get_range]
+    intro i h _
+    rw [← get_take _ _ h.1, get_range]
+    simpa using h.2
```
Comment (Komyyy) [inline Mathlib/Data/List/Iterate.lean:26]:
```suggestion
theorem take_range (m n : ℕ) : take m (range n) = range (min m n) := by
  apply List.ext_get <;> simp (config := { contextual := true }) [← get_take]
```
also, I think this theorem should be moved to `Mathlib.Data.List.Range`.

**Review by Komyyy — CHANGES_REQUESTED** [review]

File: Mathlib/Data/List/Iterate.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Data/List/Iterate.lean:26]: I think `lt_min_iff` is not available in that file

Comment (Komyyy) [reply Mathlib/Data/List/Iterate.lean:26]: Use `Nat.lt_min` in `Init.Data.Nat.MinMax` instead.

Comment (eric-wieser) [reply Mathlib/Data/List/Iterate.lean]: Nice, thanks!

**Review by Komyyy — APPROVED** [review]
Thank you!

File: Mathlib/Data/List/Range.lean [hunk 2]
```diff
@@ -84,6 +84,11 @@ theorem pairwise_le_range (n : ℕ) : Pairwise (· ≤ ·) (range n) :=
   Pairwise.imp (@le_of_lt ℕ _) (pairwise_lt_range _)
 #align list.pairwise_le_range List.pairwise_le_range
 
+theorem take_range (m n : ℕ) : take m (range n) = range (min m n) := by
```
Comment (urkud) [inline Mathlib/Data/List/Range.lean:87]: Should it be a `simp` lemma?

File: Mathlib/Data/List/Iterate.lean [hunk 3]
```diff
@@ -45,4 +45,14 @@ theorem range_map_iterate (n : ℕ) (f : α → α) (a : α) :
     (List.range n).map (f^[·] a) = List.iterate f a n := by
   apply List.ext_get <;> simp
 
+theorem iterate_add (f : α → α) (a : α) (m n : ℕ) :
+    iterate f a (m + n) = iterate f a m ++ iterate f (f^[m] a) n := by
+  induction m generalizing a with
+  | zero => simp
+  | succ n ih => rw [iterate, add_right_comm, iterate, ih, Nat.iterate, cons_append]
+
+theorem take_iterate (f : α → α) (a : α) (m n : ℕ) :
```
Comment (urkud) [inline Mathlib/Data/List/Iterate.lean:54]: Should it be a `simp` lemma?

**Review by eric-wieser — COMMENTED** [review]
@urkud: my instinct was to leave these as not `simp` lemmas, since I'm not all that familiar with what's actually useful for `simp` to do when working on lists. Presumably at some point these will end up upstream somewhere, and I'll let the upstreamers decide whether `simp` is useful.

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Pull request closed without merging**